internal/pkg/ntp/ntp.go

```diff
@@ -310,28 +324,73 @@ func (syncer *Syncer) resolveServers(ctx context.Context) ([]string, error) {
 	return serverList, nil
 }
 
-func (syncer *Syncer) queryServer(server string) (*ntp.Response, error) {
-	resp, err := syncer.NTPQuery(server)
-	if err != nil {
-		return nil, err
-	}
+func (syncer *Syncer) queryServer(server string) (*Measurement, error) {
+	if strings.Contains(server, "/dev") || strings.Contains(server, "/sys") { // sync to PTP clock
+		phc, err := os.Open(server)
 
-	syncer.logger.Debug("NTP response",
-		zap.Duration("clock_offset", resp.ClockOffset),
-		zap.Duration("rtt", resp.RTT),
-		zap.Uint8("leap", uint8(resp.Leap)),
-		zap.Uint8("stratum", resp.Stratum),
-		zap.Duration("precision", resp.Precision),
-		zap.Duration("root_delay", resp.RootDelay),
-		zap.Duration("root_dispersion", resp.RootDispersion),
-		zap.Duration("root_distance", resp.RootDistance),
-	)
+		if err != nil {
+			return nil, err
+		}
+		defer phc.Close()
 
-	if err = resp.Validate(); err != nil {
-		return resp, err
-	}
+		clockid := int32(3 | (^phc.Fd() << 3))
+
+		var ts unix.Timespec
+		err = unix.ClockGettime(clockid, &ts)
+		if err != nil {
+			return nil, err
+		}
+
+		offset := time.Until(time.Unix(ts.Sec, ts.Nsec))
+		syncer.logger.Debug("PTP clock",
+			zap.Duration("clock_offset", offset),
+			zap.Int64("sec", ts.Sec),
+			zap.Int64("nsec", ts.Nsec),
+			zap.String("device", server),
+		)
+
+		meas := &Measurement{
+			ClockOffset: offset,
+			Leap:        0,
+			Error:       err,
+			Spike:       false,
+		}
+
+		return meas, err
+	} else {
+		resp, err := syncer.NTPQuery(server)
+		if err != nil {
+			return nil, err
+		}
 
-	return resp, err
+		syncer.logger.Debug("NTP response",
+			zap.Duration("clock_offset", resp.ClockOffset),
+			zap.Duration("rtt", resp.RTT),
+			zap.Uint8("leap", uint8(resp.Leap)),
+			zap.Uint8("stratum", resp.Stratum),
+			zap.Duration("precision", resp.Precision),
+			zap.Duration("root_delay", resp.RootDelay),
+			zap.Duration("root_dispersion", resp.RootDispersion),
+			zap.Duration("root_distance", resp.RootDistance),
+		)
+
+		meas := &Measurement{
+			ClockOffset: resp.ClockOffset,
+			Leap:        resp.Leap,
+			Error:       resp.Validate(),
+			Spike:       false,
+		}
+
+		if meas.Error == nil {
+			meas.Spike = syncer.isSpike(resp)
+		}
+
+		if err = meas.Error; err != nil {
```

Comment:
why? it should return error nil, but response is invalid 

this is changing the logic

Reply:
Yes, dropped this